sensors/sensors.c

```diff
@@ -0,0 +1,630 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Sensor Manager
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/msg.h>
+#include <sys/stat.h>
+#include <sys/list.h>
+#include <sys/threads.h>
+#include <posix/utils.h>
+
+#include "sensors.h"
+
+#define CLIENT_SET_ID(id) (id + 1)
+#define CLIENT_GET_ID(id) (id - 1)
+
+#define NB_SENSOR_TYPES (sizeof(sensor_type_t) * 8)
+
+
+typedef struct {
+	sensors_ops_t ops;
+	int refs;
+	idnode_t node;
+} sensor_client_t;
+
+
+struct {
+	rbtree_t drvs;    /* registered sensors drivers */
+	idtree_t infos;   /* driver instances */
+	idtree_t clients; /* set of clients */
+
+	handle_t cLock;                          /* client's tree lock */
+	handle_t locks[NB_SENSOR_TYPES];         /* locks for each sensor type */
+	uint8_t evtNb[NB_SENSOR_TYPES];          /* number of events from all sensors */
+	sensor_event_t *events[NB_SENSOR_TYPES]; /* each row defines set of events of the same type */
+
+	uint8_t **devEvents; /* events assign to each device */
+
+	oid_t oid;
+} sensors_common;
+
+
+/* Update data from sensors */
+
+int sensors_publish(unsigned int devId, const sensor_event_t *event)
+{
+	int err = EOK;
+	uint8_t id = __builtin_ffs(event->type);
+	uint8_t evtId = sensors_common.devEvents[devId][id];
+
+	/*__builtin_ffs returns one plus the index of the least significant, otherwise 0 */
+	if (id == 0) {
+		return -EINVAL;
+	}
+
+	--id;
+	mutexLock(sensors_common.locks[id]);
+	if (id >= NB_SENSOR_TYPES || sensors_common.events[id] == NULL || sensors_common.evtNb[id] < evtId) {
+		err = -EINVAL;
+	}
+	memcpy(&sensors_common.events[id][evtId], event, sizeof(sensor_event_t));
+	mutexUnlock(sensors_common.locks[id]);
+
+	return err;
+}
+
+
+/* Handle messages */
+
+static sensor_client_t *sensors_clientFind(id_t id)
+{
+	sensor_client_t *client;
+
+	mutexLock(sensors_common.cLock);
+	client = lib_treeof(sensor_client_t, node, idtree_find(&sensors_common.clients, id));
+	if (client != NULL) {
+		client->refs++;
+	}
+	mutexUnlock(sensors_common.cLock);
+
+	return client;
+}
+
+
+static void sensors_clientPut(sensor_client_t *client)
+{
+	int refs;
+
+	mutexLock(sensors_common.cLock);
+	refs = --client->refs;
+	if (refs <= 0) {
+		idtree_remove(&sensors_common.clients, &client->node);
+	}
+	mutexUnlock(sensors_common.cLock);
+
+	if (refs <= 0) {
+		free(client);
+	}
+}
+
+
+static int sensors_open(void)
+{
+	int res;
+	sensor_client_t *client = calloc(1, sizeof(sensor_client_t));
+
+	if (client == NULL) {
+		return -ENOMEM;
+	}
+
+	mutexLock(sensors_common.cLock);
+	res = idtree_alloc(&sensors_common.clients, &client->node);
+	mutexUnlock(sensors_common.cLock);
+
+	if (res < 0) {
+		free(client);
+	}
+	else {
+		client->refs = 1;
```

Comment:
Set refs counter before adding to the tree

Reply:
done